### ISSUE TITLE
Request: On-Screen Text for the Save State Slots and Rewind in Full-Screen

### ISSUE THREAD

**Issue by xZabuzax** [punesemu/puNES#73]
Hello, if possible I would like to ask a minor feature request:

1 - Would you please provide an on-screen text in full-screen to see which Save State Slot you have selected? because we can only see the Save State Slot in windowed mode.

2 - Since you replaced "Timeline" with "Rewind" we aren't able to rewind the game in full-screen anymore, to be able to rewind the game we need to be in windowed mode and rewind it manually with the mouse. Would you please add a hotkey to be able to rewind the game in full-screen?

**Comment by punesemu** [comment]
Sorry if I answer you only now but I haven't had much free time in these weeks.
1- this already happens. When you are in the fullscreen and change slots, the corresponding slot number appears. Are you telling me that it doesn't work for you? To work you must have the "Text on screen" option in video setting enabled.
2 - You are definitely right. However, I have to invent a functional way to manage the full-screen rewind. I am open to suggestions.

**Comment by xZabuzax** [comment]
Thanks for replying, it's understandable that you're busy, don't worry about it mate, continue your work on this amazing emulator at your own pace.

1 - You're right, it is visible in full-screen but is kinda hard to see sometimes, the numbers are a bit tiny. It would be nice if the numbers where a little big bigger, it would also be nice if it had some sort of "preview" to see the picture of the game you saved in those save slots, it's an interesting feature that is only available on a few emulators. The emulator with the best implementation of this feature that I've seen so far is the Mac/Os X version of Snes9X.

2 - I've got an interesting idea of how you could implement the current "rewind" in full-screen. You could add a hotkey which will open the "Rewind Buttons" feature in the corner of the screen or something, this will pause the emulator and you could use "Rewind Buttons" with the mouse (like you currently do in windowed mode) and you could also use it with the keyboard using the arrow keys. For example:

Pressing "Left" will rewind it by 1 frame, pressing "Right" will forward it by 1 frame.
Pressing "Down" will rewind it automatically, pressing "Up" will forward it automatically.
Pressing "Space Bar" will serve as Play/Pause

In other words, that "Rewind Buttons" feature will be similar like the media players that we use when watching movies because in the media players you could rewind, forward, pause/play with the keyboard as well with either the arrow keys and the space bar.

What do you think of that idea?

**Comment by punesemu** [comment]
I like the idea, I will implement it.

**Comment by nice-useles-person** [comment]
The goal of such features is usually ease of access. So, the first priority should be to implement these features in a way that a person using controller should be able to use this whole functionality using controller buttons only and also keyboard buttons if the user is out of buttons on his controllers. Here is my humble suggestion:
*A configurable key (on controller as well as keyboard) might be used to activate Rewind Mode(In Full Screen as well as window mode) that will pause  emulation and display an overlay or any other effect to indicate that rewind mode is active. (I think it will be very cool, If an effect or shader is be applied in rewind mode like the one that happens when we used to rewind or forward VHS. Those distortion lines on TV, That will be very nostalgic)
*A separate Button Configuration can be added that is only valid when Rewind Mode is active.
*Once, the rewind mode is active, all rewind functions can be mapped to all controller or keyboard buttons
*For example, I use an Xinput controller and will prefer to use D-PAD right and left  for forwarding or reversing frames manually. up and down for automatic forward or reverse. and A for Pause and X (or same key for activating rewind mode) for Play. As soon as it plays, The rewind mode will automatically exit of course
*You will have to add another tab in settings for rewind keys configuration just like normal game-play keys configuration. The key for activating rewind mode should however be added in hot keys section

Hopefully, It will make user experience very pleasant.

**Comment by nice-useles-person** [comment]
I know it is not important for functionality. But it will be very cool to have this kind of overlay when using Rewind.
https://i.ytimg.com/vi/zByO2TmM1WU/maxresdefault.jpg

**Comment by punesemu** [comment]
Could you try this [version](https://www.dropbox.com/s/hrsdy1c2na6y0zl/punes.exe?dl=0)? I introduced the shortcuts to activate and deactivate the "rewind mode" and the shortcuts for play, pause, step backward, step forward, fast backward and fast forward, all programmable. Obviously you can also associate the keys that are used for the movements or for the buttons A and B.

**Comment by nice-useles-person** [comment]
Tested this functionality and it is working fine as expected. I just wish the overlay used more sophisticated icons for displaying current operations and an effect while rewind mode is active. Also if an option is implemented in settings to adjust size/scale of overlay objects like texts or any icon, that will be awesome.

**Comment by xZabuzax** [comment]
> Could you try this [version](https://www.dropbox.com/s/hrsdy1c2na6y0zl/punes.exe?dl=0)? I introduced the shortcuts to activate and deactivate the "rewind mode" and the shortcuts for play, pause, step backward, step forward, fast backward and fast forward, all programmable. Obviously you can also associate the keys that are used for the movements or for the buttons A and B.

The link is not working anymore. Could you please provide a new link? I want to test the new functionality of the rewind feature.

**Comment by punesemu** [comment]
Try the latest [WIP](https://github.com/punesemu/puNES#wip), it's updated to the last commit that already contain the management of shortcuts. I just sent the last two commits (https://github.com/punesemu/puNES/commit/090de95a28f6b83469f4872728b661b71a6bee6d and https://github.com/punesemu/puNES/commit/a82094c2315d6d07fd1e507e0938316282a6d6b7) that improve their management.

**Comment by xZabuzax** [comment]
Just tested the new mechanic of the rewind feature and it is working as intended, this is exactly what I described in my idea above, pretty cool.
I liked how pressing the rewind or forward key more than once speeds up the rewind and forwarding process, that's pretty cool as well.

I'm very pleased with how the rewind feature turned out to be, this is pretty original in an emulator, hopefully the other emulator developers learn from this and implement this idea as well in their emulators.

It would be pretty cool having some sort of visible effect when the rewind feature is enabled, it can be a feature than can be enabled or disabled, this isn't necessary but it would be cool regardless, some users like the simplicity of an emulator and that's fine but other users also likes an "eye-candy" emulator with features like that. Whether you implement the "eye-candy" part is up to you though, personally I'm happy how the rewind feature turned out to be.

**Comment by nice-useles-person** [comment]
> Just tested the new mechanic of the rewind feature and it is working as intended, this is exactly what I described in my idea above, pretty cool.
> I liked how pressing the rewind or forward key more than once speeds up the rewind and forwarding process, that's pretty cool as well.
> 
> I'm very pleased with how the rewind feature turned out to be, this is pretty original in an emulator, hopefully the other emulator developers learn from this and implement this idea as well in their emulators.
> 
> It would be pretty cool having some sort of visible effect when the rewind feature is enabled, it can be a feature than can be enabled or disabled, this isn't necessary but it would be cool regardless, some users like the simplicity of an emulator and that's fine but other users also likes an "eye-candy" emulator with features like that. Whether you implement the "eye-candy" part is up to you though, personally I'm happy with how the rewind feature turned out to be.

I wanted to say the same things you have said. It will feel amazing if if felt like rewinding on a system from those NES days like VHS. I have edited some screenshots to show what I mean.

![Rew](https://user-images.githubusercontent.com/26537640/61585822-e522c180-ab7e-11e9-843d-0b8f8c243575.jpg)
![Fwd](https://user-images.githubusercontent.com/26537640/61585824-eeac2980-ab7e-11e9-92e9-1d0f679decfe.jpg)
![Pause](https://user-images.githubusercontent.com/26537640/61585825-f10e8380-ab7e-11e9-8520-1ad822aa73b8.jpg)
![Play](https://user-images.githubusercontent.com/26537640/61585826-f2d84700-ab7e-11e9-98b3-1c76f9d8d0fa.jpg)

The animated distortion may be applied as soon as rewind mode is active and the icon may be displayed to convey current operation. The speed of distortion animation may convey rewind speed multiplier. Users may be given options in setting to adjust scale of overlay text and icons, and amount of distortion effect.

A very good game called hotline Miami can be used a good reference fro implementation of such effects. Here is an example from the game with amazing sound effects to go along visual effects.

https://youtu.be/Ikxonj8tkbI

I personally love this emulator and I want it to be as enjoyable as possible. Therefore I keep suggesting these things. I apologize in advance to the creator if I ever seems demanding or ungrateful.

**Comment by xZabuzax** [comment]
@nice-useles-person

Yeah that's what I'm talking about, I would also love to see those effects when "rewind mode" is on. Personally I'm more of an "eye-candy" type of guy when it comes to emulator, I love emulators that I can configure the way I like to (with hotkeys and all) and also love emulators with plenty of features or visual effects.

In my opinion puNES is doing a great job with the configurable stuff and the eye-candy parts and that's one of the reasons why puNES became my default NES emulator. Having those extra features with the effects in "rewind mode" will only increase my pleasure even more, they are not necessary but they are still going to increase my pleasure with the emulator.

In the end is up to the developer to add those features, I'm not in a hurry though, he can take his time to think about it. If he decides to not add those extra features then I won't mind, I'm still happy with the emulator the way it is.

**Comment by punesemu** [comment]
I like the idea of adding a visual effect during rewind and, free time permitting, I think I will do it in the next release (I also plan to review the overlay management) but for now I can release the 0.105 version that was supposed to be a bugfix release with all the bugs I have been corrected in these 3 months plus the management of shortcuts in the rewind. Thank you so much for your support.

**Comment by xZabuzax** [comment]
> I like the idea of adding a visual effect during rewind and, free time permitting, I think I will do it in the next release (I also plan to review the overlay management) but for now I can release the 0.105 version that was supposed to be a bugfix release with all the bugs I have been corrected in these 3 months plus the management of shortcuts in the rewind. Thank you so much for your support.

Awesome!, I'm pretty sure there's going to be a lot of users that are going to be pleased with that extra feature as well. Personally I believe is a good idea to add that feature because your "rewind mode" is very original on emulators and being original and all it deserves some love with the aesthetics.

You redefined the way emulators should use the rewind functionality, I just wish the other emulator authors implement your idea in their emulators.
